pkg/component/shared/kubeapiserver.go

```diff
@@ -220,7 +220,15 @@ func DeployKubeAPIServer(
 
 	kubeAPIServer.SetAutoscalingReplicas(computeKubeAPIServerReplicas(values.Autoscaling, deployment, wantScaleDown))
 
-	if deployment != nil && values.Autoscaling.HVPAEnabled {
+	// For safety reasons, when the Deployment exists we don't overwrite the kube-apiserver container resources
+	// although it is not required in all cases. Few cases that require it:
+	// - When autoscaling mode is HVPA. hvpa-controller updates the resources in Deployment spec directly.
+	//   Overwriting the resources in the Deployment spec would revert hvpa-controller's recommendations.
+	// - When scale-down is disabled, operators might want to overwrite the kube-apiserver container resource requests.
+	// - When transitioning from HVPA to HPAAndVPA autoscaling mode, we need to preserve the kube-apiserver container resources
+	//   to do not cause an unwanted rollout that might be breaking. Otherwise, we would scale down from the potentially
+	//   high resource requests (set by HVPA) to the initial resource requests in HPAAndVPA mode.
+	if deployment != nil {
```

Comment:
> HVPA actions are based on VPA recommendations. So the only scenario in which we would really revert an existing kube-apiserver to initial resource requests would be

Yes, you are right and I agree. The last sentence in the last bullet is quite correct:
> Otherwise, we would scale down from the potentially high resource requests (set by HVPA) to the initial resource requests in HPAAndVPA mode.

IIUC, if we don't preserve the kube-apiserver container resource requests in the Deployment during the HVPA -> VPAAndHPA transition, maybe we would only revert to the initial resource requests if the vpa-recommender gets restarted and we lose the historical data that is somehow kept in memory in the vpa-recommender. I really don't want to cause a rollout of the Deployment and to look into very details when the Deployment gets updated, when the VPA gets created, will always the vpa-webhook mutate the new Pods from the rollout, etc.

So, for safety reasons, I think we still need to do NOT revert the kube-apiserver container resource requests during the HVPA -> VPAAndHPA transition because:
- this would cause and unnecessary rollout always because Deployment spec is changed -> new ReplicaSet is created
- there is a corner case that is described above that the VPA's webhook might not be able to mutate and might cause revert to the initial resource requests.
- IIUC, one of the main drawbacks of the HVPA is that it does horizontal scaling to maxReplicas first and only this it does vertical scaling. Which makes me think that we have a lot of VPA checkpoints (VPA recommendations) that are not applied by HVPA because of the above limitation/drawback. Hence, a rollout always during the HVPA -> VPAAndHPA would always mean applying the VPA recommendations (that were ignored so far due to the HVPA limitation) right away.

IMO, the kube-apiserver container resource requests should not be mutated when:
- HVPA is enabled
- scaled-down is disabled
- and we are in the transition from HVPA -> VPAAndHPA

Let me also cc @vlerenc as we were discussing this topic with him offline and we agreed to do not try to cover every case in the if statement but to never mutate/overwrite the kube-apiserver resource requests in the Deployment.

Reply:
I reworked the last bullet to:
```go
	// - When transitioning from HVPA to VPAAndHPA autoscaling mode, we need to preserve the kube-apiserver container resources
	//   to do not cause an unwanted Deployment rollout by overwriting the container resources set by the HVPA with the initial
	//   resources in VPAAndHPA mode.
```

Reply:
My two cents are basically:
- Changing the requests (setting them to the initial values) will definitely (and needlessly) roll the KAPIs, even if VPA will mutate them in the very same way as before
- There is the very remote chance of the problem identified and described above
- We may not know of all such corner cases, so there is always a risk when you touch something and we really don't want to touch too much if we don't have to
- If we detect some form of related time bomb to these requests some hours after deployment, we may still have time to defuse it, if we haven't rolled the deployments yet in their respective maintenance time windows
- I have no good reason to change/reset to the initial requests (asking the question in reverse: why would I want to reset to the initial requests?); if anything, the current requests make more sense for this particular deployment (more than the general initial requests)

Reply:
If I understand this correctly, this would mean that we will never get rid of this piece of code, which was only introduced because that's how HVPA worked. To me, this is harder to understand and keep in mind as context for the future than clearly scoping this method to the single purpose of not resetting snowflakes.
Do we already have a dedicated test documenting the desired behavior for snowflakes? Just to make sure that future versions of us don't go in and just remove this method.

Reply:
I can add a TODO to adapt this check when we drop the HVPA autoscaling mode (if we have success with VPAAndHPA mode). Then the check would look like:
```go
	if deployment != nil && values.Autoscaling.ScaleDownDisabled {
```

I.e preserve the kube-apisever container resources only when the Deployment exists and scale down is disabled.

Is everybody fine with such TODO and with such change in the future?

---

> Do we already have a dedicated test documenting the desired behavior for snowflakes?

There is unit test verifying the kube-apiserver container resources are passed as autoscaling values unconditionally when the Deployment exists.
https://github.com/gardener/gardener/blob/946789d3ad2327cbe96ff3c3cb612e7549d6e8e3/pkg/component/shared/kubeapiserver_test.go#L951-L1005

Reply:
I don't think you'll be able to change the check as proposed, the same reasoning as above still applies: this will cause a rollout of the KAPIs. For all non snowflake clusters, this would still roll the Deployment, right?

For the test case you referenced: I was thinking about explicitly enforcing that the KAPI resources for clusters marked with `alpha.control-plane.scaling.shoot.gardener.cloud/scale-down-disabled` are not overwritten – as some sort of documentation and safety guard of why this function is still here, even when at some point in the future the HVPA references will be gone.

Reply:
> I don't think you'll be able to change the check as proposed, the same reasoning as above still applies: this will cause a rollout of the KAPIs for all non snowflake clusters, right?

Good point. I don't think I have a good answer to it. The question is whether such a rollout in the future (when HVPA is removed and VPAAndHPA is the used autoscaling mode) would cause issues for snowflake clusters. My naive expectation would be "no" expecting that VPA will mutate the Pods and apply its recommendations.
The other question raised by Vedran is "do we need to do it (this unnecessary rollout)?"

But I am open to add such TODO if you don't see problems with it.

---

> For the test case you referenced: I was thinking about explicitly enforcing that the KAPI resources for clusters marked with `alpha.control-plane.scaling.shoot.gardener.cloud/scale-down-disabled` are not overwritten – as some sort of documentation and safety guard of why this function is still here, even when at some point in the future the HVPA references will be gone.

It is a little bit writing the test against to the doc string (the concrete cases) and not against the implementation (the general one). Anyways, I adapted the test cases to cover the bullet points instead of covering actual implementation.